### DISCUSSION TITLE
Updated MachineState.context is not available within the machineService "onDone" listener, in a React Functional Component (when a final state is reached)

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2270] [Q&A]
Hello! 
I have a React Functional Hooks component that looks like this. It makes use of `@xstate/react` package & employs `xstate` via the `useMachine` hook.

```js
export const MyComponent = React.memo((props) => {
   const [machineState, dispatchToMachine, machineService] = useMachine(createMyMachine(props));

   // empty deps array to emulate componentDidMount
   useEffect(() => {
        dispatchToMachine({type: "SOME_EVT", payload: {}});
        machineService.onDone(() => {
           // do something with machineState.context.<someValue>
        });
   }, []);

   return <div>/* some JSX here */</div>
});
```

This is the structure of my machine. My component is wrapped within `React.memo`.
Using the `useEffect` hook to emulate `componentDidMount`, I dispatch a "SOME_EVT" to the machine to get things started. The machine does a lot of work and finally reaches one of the **"final"** states.

I have hooked up an `onDone` handler via the `machineService`, to do something once the machine reaches one of its **"final"** states. 
_But, the `machineState.context`, when accessed over here, still retains only the old initial Value of context. All the assign actions done to update the context is not reflected here._

Q) Is this a right way to setup an `onDone` listener. I set it up this way so that its a one-time thing. But this does not work.
What would be the right place to wire up an `onDone` handler so that I'm able to get access to the latest updated context ?
Or, what is the best way to execute some code once the machine has reached one of its **"final"** states

**Comment by dsathyakumar** [comment]
Consequently, this makes me wonder about `machineService.subscribe`, which seems to get a `state` parameter within the listener, and, this `state` has access to the updated context. Also, if `machineService.subscribe` is to be wired on, what would be the right place for it ? Surely, don't want to be wiring a new listener every time `machineService` or `machineState` changes 🤔

**Reply by Andarist** [reply]
It would be best if you could model everything within the machine - without resorting to `subscribe`. What's your use case? What are you trying to do?

**Comment by Andarist** [comment]
> But, the machineState.context, when accessed over here, still retains only the old initial Value of context. All the assign actions done to update the context is not reflected here.

This is because `machineState` is immutable and because how React rerendering works you only get access to the very first closure variables there - from the time when the machine has been instantiated and not from the time when `onDone` has been called.

> Or, what is the best way to execute some code once the machine has reached one of its "final" states

add `exit` handler on the top of your machine

**Comment by dsathyakumar** [comment]
Thanks for the response @Andarist 👍 
Yes it was a case of stale closures. Although, I was able to get around it by sending the needed latest context, via the "data" prop of the done event. 
So, in the final state I did something like
```js
data: (ctx, evt) => ctx
```

**Comment by dsathyakumar** [comment]
Although my question was, where inside a react functional component would this "wiring" of `onDone` or `subscribe` or `onStop` be done ?
